_posts/2024-10-24-autorebalancing-on-scaling.md

```diff
@@ -0,0 +1,150 @@
+---
+layout: post
+title:  "Autorebalancing on cluster scaling"
+date: 2024-10-24
+author: paolo_patierno
+---
+
+Scaling up or down an Apache Kafka cluster is usually done in order to achieve different goals and to get several benefits.
+When scaling up, by adding new brokers, the cluster is capable to handle more load in terms of data transmitted across clients.
+When scaling down, by removing some brokers from the cluster, we are able to reduce costs and energy consumption when there is not that much demand for performance due to a lower load.
+But scaling the cluster makes more sense if it's done automatically, for example by leveraging the HPA (Horizontal Pod Autoscaler) within Kubernetes.
+The problem comes when, after scaling up, the cluster is not balanced and the newly added brokers are just empty and there is the need to rebalance topics' partitions to equally spread the load across all the brokers (older and newer).
+On the other side, scaling down could be not possible when brokers to be removed are hosting topics' partitions, unless we accept to have offline replicas and losing HA (High Availability).
+In both scenarios, rebalancing the cluster by using the Cruise Control integration within Strimzi is the solution.
+Doing it automatically is even better and this is what we are going to talk about in this blog post.
+
+<!--more-->
+
+### Scaling and rebalancing, how do you do today?
+
+When scaling up an Apache Kafka cluster, the newly added brokers don't get any partitions of the already existing topics hosted on the other brokers.
+The new brokers are just empty and they are going to get partitions for newly created topics based on the distribution mechanism within Apache Kafka itself.
+Quite often, this is not what you want because you are scaling up in order to spread the load across more brokers and getting better performance.
+Today, the only way to do so with Strimzi, is about running a manual rebalancing by leveraging the `KafkaRebalance` custom resource with the `add-brokers` mode (listing the added brokers) and the goals to be satisfied by Cruise Control.
+This way, some of the already existing topics' partitions are moved from the old brokers to the new ones, making the cluster more balanced.
+
+When scaling down, the operation of removing brokers could be blocked by the Strimzi operator because those brokers are hosting topics' partitions and the immediate shutdown will cause offline replicas and losing HA.
```

Comment:
```suggestion
When scaling down, the operation to remove brokers could be blocked by the Cluster Operator because those brokers are hosting topic partitions and the immediate shutdown will cause offline replicas and loss of HA.
```